Finish P0768R1 by adding Spaceship CPOs

StephanTLavavej opened the pull request:
Resolves #64 by implementing WG21-N4861's spaceship customization point objects specified in [[cmp.alg]](https://eel.is/c++draft/cmp.alg).

Note that this is targeted at `master`, as it doesn't directly interact with #62 which is being developed in `feature/spaceship`.

* `<charconv>`
  + Lift `_Floating_type_traits` into a central header, `<type_traits>`. It was the only thing using `<float.h>` here (for `FLT_MANT_DIG` etc.).
* `<type_traits>`
  + Include `<cstdint>` needed by `_Floating_type_traits`. This is a *very* small header, "core" compatible, and it's actually surprising that `<type_traits>` wasn't already dragging it in.
  + Move `_Floating_type_traits` here. It's functionally unchanged, except that I've eliminated the `<float.h>` dependency by directly writing out the literal values. I believe that this makes the masks easier to understand, and I've retained the original computations as comments.
  + This also adds `_Floating_type_traits<long double>`, needed now but not by `<charconv>`.
  + `<charconv>` includes `<xutility>` includes `<utility>` includes `<type_traits>`.
  + `<compare>` includes (in concepts mode) `<concepts>` includes `<type_traits>`.
* `<xutility>`
  + Lift `std::ranges::_Choice_t` and `std::_Implicitly_convertible_to` into a central header, `<concepts>`. Both will remain guarded by `__cpp_lib_concepts`. `_Choice_t` is additionally being lifted out to `std`.
  + `<xutility>` includes `<utility>` includes (in concepts mode) `<concepts>`.
* `<compare>`
  + Include `<bit>` for `bit_cast`. It's a "core" header, although it does drag in `<limits>`.
  + **Possible alternative:** We could move `_Bit_cast` in `<xutility>` up to `<type_traits>`, except that it wants `memcpy` for `__CUDACC__` and `<type_traits>` doesn't include `<cstring>`. Maybe we could write a raw byte-copying loop?
  + Implement the 6 CPOs.
  + I encountered 2 issues in the Standardese, which @CaseyCarter and I consider to be defects. I filed #1374 to track this. They have essentially zero impact for users (the "passing arguments as lvalues" is extremely difficult to observe, and is consistent with more modern Standardese elsewhere; the "block unqualified name lookup" technique is nearly impossible to observe, as I understand it, but should be clearly stated by the Standard.)
  + This "directly" implements the Standardese, using @CaseyCarter's clever `_Choice`/`_Strategy` system.
  + This implements strong and weak ordering for floating-point numbers by directly inspecting their representations; I have carefully commented each step, *and* written comprehensive tests.
* `<yvals_core.h>`
  + P0768R1 is now fully implemented; remove the "partially implemented" comment.
  + Define the feature-test macro, guarded by `__cpp_lib_concepts`. Note that `__cpp_lib_three_way_comparison` is being defined to the lower value `201711L` for P0768R1; it will be increased to the larger value `201907L` for P1614R2.
* `tests/libcxx/expected_results.txt`
* `tests/libcxx/skipped_tests.txt`
* `tests/std/tests/VSO_0157762_feature_test_macros/test.cpp`
  + Test the feature-test macro.
* `tests/std/test.lst`
* `tests/std/tests/P0768R1_spaceship_cpos/env.lst`
* `tests/std/tests/P0768R1_spaceship_cpos/test.cpp`
  + Add comprehensive tests (requiring the `concepts_matrix`, but we don't need the *strict* concepts matrix - I was able to resolve all of the issues with `/permissive` mode).
  + Verifying "expression-equivalent" is a lot of work! I'm testing the return type, exception specification, and compile-time/run-time behavior. Because there are so many aspects to test, 6 different CPOs, and so many different cases that they activate, I've commented each test with the Standardese bullet point that it's testing.
  + In `test_floating()`, I've devised a system to test every possible "interesting" pair of floating-point values.

miscco commented on stl/inc/xutility:
```diff
@@ -644,13 +644,6 @@ struct iterator_traits<_Ty*> {
 
 // CUSTOMIZATION POINT OBJECT ranges::iter_move
 namespace ranges {
-    // STRUCT TEMPLATE _Choice_t
```
To be safe here, should we explicitly include `<compare>`? It is not immediately clear how those definitions end up here

StephanTLavavej replied:
We occasionally explicitly include a header even if we get it through another path, and sometimes we comment where a piece of machinery lives. Looking at this again, I think that it's just weird for these helper templates to live in `<compare>` even though it happens to be the first place where they're used. As the throughput implications are near-zero and we'll probably need them elsewhere in the future, I will promote them all the way up to `<concepts>` (as they are concept helpers) and that way I don't think we need either explicit includes or comments.

miscco commented on stl/inc/compare:
```diff
@@ -340,9 +341,503 @@ struct compare_three_way {
+
+    template <class _Ty1, class _Ty2>
+    concept _Has_ADL = requires(_Ty1&& _Left, _Ty2&& _Right) {
+        static_cast<strong_ordering>(strong_order(_Left, _Right));
```
I am wondering why this is using a cast versus an explicit constraint on the return type
```cpp
        { strong_order(_Left, _Right) } -> strong_ordering;

```

Applies below

miscco replied:
Argh, forgot to remove this one after reading the standard

StephanTLavavej replied:
There is a related question though - the Standardese asks whether `strong_­ordering(strong_­order(E, F))` is well-formed. I converted this to a `static_cast` because I dislike functional-style casts; however, the difference may be observable.

CaseyCarter replied:
> the difference may be observable.

Is this true when the target type is a non-reference class type that we control? My imagination is lacking today.

StephanTLavavej replied:
I will resolve this as we can't imagine a way to observe any difference from the Standardese. (For the record, as @miscco realized, explicitly constraining the return type would prevent the function call from returning something not `strong_ordering` but convertible to `strong_ordering`; the functional-style cast and `static_cast` permit that.)

miscco commented on stl/inc/xutility:
```diff
@@ -644,13 +644,6 @@ struct iterator_traits<_Ty*> {
 
 // CUSTOMIZATION POINT OBJECT ranges::iter_move
 namespace ranges {
-    // STRUCT TEMPLATE _Choice_t
```
How does this gets included back into `<xutility>`

Should there be an explicit include of `<compare>`?

StephanTLavavej replied:
Resolving duplicate comment.

miscco commented on stl/inc/compare:
```diff
@@ -340,9 +341,503 @@ struct compare_three_way {
+                } else {
+                    return _Left_uint <=> _Right_uint;
+                }
+            } else {
```
I think this should be an `else if`. The `static_assert` below kind of obscures that we should never reach the `else`, which is quite clear via the `static_assert(_Always_false<_Ty>, "should be unreachable");` that is use in the other CPOs

StephanTLavavej replied:
Thanks for catching this - I was trying to simplify the original pattern, but obscured it, *and* introduced inconsistency between files while doing so. Changed all occurrences to use `_Always_false` as you recommended.

miscco reviewed (commented):
Looks great, I had a tiny question about the include chain and another one regarding readability

CaseyCarter commented on tests/std/tests/P0768R1_spaceship_cpos/test.cpp:
```diff
@@ -0,0 +1,467 @@
+};
+
+template <const auto& CPO, typename E, typename F>
+struct CpoResultImpl<CPO, E, F, void_t<decltype(CPO(declval<E>(), declval<F>()))>> {
```
Peculiar (but not wrong) that you're using `void_t` here instead of a requires-clause for the partial specialization when the test is using concepts_matrix.

StephanTLavavej replied:
Peculiar (but not wrong) that you're recrystallizing dilithium with gamma-ray photons instead of a theta-matrix compositor, like any modern starship built by Utopia Planitia in the 2360s.

(Silly me using archaic future tech)

CaseyCarter commented on stl/inc/compare:
```diff
@@ -340,9 +341,503 @@ struct compare_three_way {
+                return static_cast<weak_ordering>(compare_three_way{}(_Left, _Right));
+            } else {
+                static_assert(_Strat == _St::_Strong);
+                return static_cast<weak_ordering>(_STD strong_order(_Left, _Right));
```
Since only ADL cases get here (per my comment on the concept) we could add a `void strong_order();` block and strike `_STD` here to dispatch directly.

StephanTLavavej replied:
Done with an `/* ADL, throughput optimization */` comment. Also done for the `noexcept` expression in `_Choose()`.

StephanTLavavej commented on tests/std/tests/P0768R1_spaceship_cpos/test.cpp:
```diff
@@ -0,0 +1,467 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// Test the customization point objects in N4861 [cmp.alg].
```
Thanks, didn't know about that test. Done.

StephanTLavavej commented on stl/inc/compare:
```diff
@@ -340,9 +341,503 @@ struct compare_three_way {
+    bool _No_throw = false;
+};
+
+// Note: The following CPOs are passing arguments as lvalues; LWG issue not yet submitted as of 2020-10-09.
```
Filed #1374 and assigned to you. I'll update these comments to cite it.

StephanTLavavej commented on stl/inc/compare:
```diff
@@ -340,9 +341,503 @@ struct compare_three_way {
+    void strong_order(); // Block unqualified name lookup; LWG issue not yet submitted as of 2020-10-09.
+
+    template <class _Ty1, class _Ty2>
+    concept _Has_ADL = requires(_Ty1&& _Left, _Ty2&& _Right) {
```
Great catch - this is a relic of when I was attempting to perfect-forward everything. Changed all to lvalue references except for the function call operators themselves.

StephanTLavavej commented on stl/inc/compare:
```diff
@@ -340,9 +341,503 @@ struct compare_three_way {
+    public:
+        // clang-format off
+        template <class _Ty1, class _Ty2>
+            requires (_Choice<_Ty1, _Ty2>._Strategy != _St::_None)
```
Excellent - I never would have noticed that. Changed all.

StephanTLavavej commented on stl/inc/compare:
```diff
@@ -340,9 +341,503 @@ struct compare_three_way {
+
+                // 3. Fast path: any negative value is less than any positive value.
+                if (_Left_negative != _Right_negative) {
+                    return _Left_negative ? strong_ordering::less : strong_ordering::greater;
```
I believe that's far too clever - it has to reverse right/left because "negative" being "true" being "1" is the opposite of what we want. As written, the single line of code matches the comment. No change made here.

StephanTLavavej commented on stl/inc/compare:
```diff
@@ -340,9 +341,503 @@ struct compare_three_way {
+
+    template <class _Ty1, class _Ty2>
+    concept _Can_strong_order = requires(_Ty1&& _Left, _Ty2&& _Right) {
+        static_cast<weak_ordering>(_STD strong_order(_Left, _Right));
```
Yep - I double-checked your analysis and found no issues. (Not even with the `weak_ordering` versus `strong_ordering` return type variation; if we can't `compare_three_way` to get `weak_ordering`, we certainly can't get a `strong_ordering` out of it - although the reverse would be possible.)

Done with a "throughput optimization" comment.

StephanTLavavej commented on stl/inc/compare:
```diff
@@ -340,9 +341,503 @@ struct compare_three_way {
+
+    template <class _Ty1, class _Ty2>
+    concept _Can_weak_order = requires(_Ty1&& _Left, _Ty2&& _Right) {
+        static_cast<partial_ordering>(_STD weak_order(_Left, _Right));
```
I've implemented a complete throughput optimization along these lines.

StephanTLavavej commented:
@miscco, @CaseyCarter - I've addressed your feedback with:

* Update general CPO properties test.
* Filed GH issue for wording cleanups.
* Take lvalue references in helper concepts.
* Move helpers up to `<concepts>`.
* Use requires instead of `void_t`.
* Use `_Choice<_Ty1&, _Ty2&>` to reduce template instantiations.
* Use `else if constexpr` followed by `_Always_false` for consistency.
* Optimize throughput by directly invoking ADL.

I also added a couple of additional changes:

* Extract `_Can_fallback_eq_lt`, rename `_Can_fallback_eq_lt_twice`.
* Comment ADL usage.

StephanTLavavej commented on stl/inc/compare:
```diff
@@ -340,9 +341,496 @@ struct compare_three_way {
+            } else if constexpr (_Strat == _St::_Three) {
+                return static_cast<weak_ordering>(compare_three_way{}(_Left, _Right));
+            } else if constexpr (_Strat == _St::_Strong) {
+                return static_cast<weak_ordering>(/* ADL, throughput optimization */ strong_order(_Left, _Right));
```
Fixed. For extreme thoroughness, I've added test coverage, which fails without the fix:

```
S:\GitHub\STL\out\build\x86\out\inc\compare(478): error C2280: 'TestUdt::WeirdOrdering<std::strong_ordering,true>::operator AnythingElse(void) const<std::weak_ordering>': attempting to reference a deleted function
        with
        [
            AnythingElse=std::weak_ordering
        ]
S:\GitHub\STL\tests\std\tests\P0768R1_spaceship_cpos\test.cpp(120): note: see declaration of 'TestUdt::WeirdOrdering<std::strong_ordering,true>::operator AnythingElse'
        with
        [
            AnythingElse=std::weak_ordering
        ]
S:\GitHub\STL\out\build\x86\out\inc\compare(485): note: see reference to function template instantiation 'std::_Choice_t<std::_Weak_order::_Cpo::_St> std::_Weak_order::_Cpo::_Choose<_Ty1,_Ty2>(void) noexcept' being compiled
        with
        [
            _Ty1=TestUdt::StrongWeird<true> &,
            _Ty2=TestUdt::StrongWeird<true> &
        ]
S:\GitHub\STL\out\build\x86\out\inc\compare(490): note: see reference to variable template 'const std::_Choice_t<std::_Weak_order::_Cpo::_St> std::_Weak_order::_Cpo::_Choice<TestUdt::StrongWeird<1> &,TestUdt::StrongWeird<1> &>' being compiled
S:\GitHub\STL\tests\std\tests\P0768R1_spaceship_cpos\test.cpp(275): note: see reference to alias template instantiation 'CpoResult<std::_Cpos::weak_order,TestUdt::StrongWeird<true>,E>' being compiled
        with
        [
            E=TestUdt::StrongWeird<true>
        ]
```

The test coverage additionally verifies whether the conversion from the user-defined result type to the Standard comparison category type can influence the `noexcept`ness of the expression.

statementreply commented on tests/std/tests/P0768R1_spaceship_cpos/test.cpp:
```diff
@@ -0,0 +1,554 @@
+    if constexpr (is_same_v<Floating, float>) {
+        const pair<int, float> rank_value_pairs[]{
+            {10, bit_cast<float>(0xFFC01234u)}, // negative quiet NaN, payload bits
+            {10, bit_cast<float>(0xFFC00000u)}, // negative quiet NaN, no payload bits
+            {20, -numeric_limits<float>::infinity()}, // negative infinity
```
Do we need test coverage for signaling NaNs?

Note that it could be currently impossible to generate a `float` signaling NaN on the MSVC compiler at compile time. It emits a quiet NaN for `bit_cast<float>(0xFF800001u)` or `numeric_limits<float>::signaling_NaN()`.

Same below.

StephanTLavavej replied:
Ideally, yes, this would test signaling NaNs - but as you noted, the compiler bugs make it difficult. I think we can live without test coverage for them, given that the code is clearly written to handle them uniformly (the main question is whether I got the "infinity plus one" logic correct, but I think that can be verified by inspecting the code).

statementreply commented on tests/std/tests/P0768R1_spaceship_cpos/test.cpp:
```diff
@@ -0,0 +1,554 @@
+template <typename Floating>
+constexpr void test_floating() {
+    if constexpr (is_same_v<Floating, float>) {
+        const pair<int, float> rank_value_pairs[]{
+            {10, bit_cast<float>(0xFFC01234u)}, // negative quiet NaN, payload bits
```
Do we need test coverage for the limiting case (`0xFFFFFFFFu`)?

Same below.

StephanTLavavej replied:
Good idea, I'll add that!

StephanTLavavej commented:
/azp run
